File: code/Kotlin/libs/annotation_processing/apt/compiler/src/main/java/inject/processor/creator/CommonMethod.java
package inject.processor.creator;

import static javax.lang.model.element.Modifier.PRIVATE;
import static javax.lang.model.element.Modifier.PUBLIC;
import static javax.lang.model.element.Modifier.STATIC;

import com.squareup.javapoet.ClassName;
import com.squareup.javapoet.MethodSpec;
import com.squareup.javapoet.TypeName;

import java.util.ArrayList;
import java.util.List;

import javax.lang.model.element.Element;

import inject.classname.AndroidClassName;
import inject.classname.OtherClassName;
import inject.classname.PrimitiveClassName;

/**
 * @author yuansui
 * @since 2020/9/8
 */
class CommonMethod extends BaseMethod {

    CommonMethod(CreatorProcessor processor, Element annotatedElement) {
        super(processor, annotatedElement);
    }

    MethodSpec constructor() {
        return MethodSpec.constructorBuilder()
                .addModifiers(PRIVATE)
                .build();
    }

    MethodSpec create(String fileName) {
        ClassName clzName = ClassName.get(processor.getPackageName(annotatedElement), fileName);

        /*
          生成create方法
         */
        String simpleName = clzName.simpleName();
        MethodSpec.Builder builder = MethodSpec.methodBuilder("create")
                .addModifiers(PUBLIC, STATIC)
                .addStatement(simpleName + " builder = new " + simpleName + "()")
                .returns(clzName);

        // 添加必须的参数
        for (Element e : requireFields) {
            String paramName = processor.getParamName(e);
            builder.addParameter(processor.createNullableParam(e, paramName));
            builder.addStatement("builder.$N = $N", paramName, paramName);
        }
        builder.addStatement("return builder");

        return builder.build();
    }

    /**
     * 根据optional生成链式调用的方法
     */
    List<MethodSpec> optionalSetters(String fileName) {
        List<MethodSpec> specs = new ArrayList<>();

        for (Element e : optFields) {
            String paramName = processor.getParamName(e);
            MethodSpec method = MethodSpec.methodBuilder(paramName)
                    .addModifiers(PUBLIC)
                    .addParameter(processor.createNullableParam(e, paramName))
//                    .addParameter(getTypeNameBox(e), paramName)
                    .addStatement("this.$N = $N", paramName, paramName)
                    .addStatement("return this")
                    .returns(ClassName.get(processor.getPackageName(annotatedElement), fileName))
                    .build();

            specs.add(method);
        }

        return specs;
    }

    MethodSpec newIntent() {
        MethodSpec.Builder builder = MethodSpec.methodBuilder("newIntent")
                .addModifiers(PUBLIC, STATIC)
                .addParameter(processor.createNonNullContext())
                .addStatement("$T intent = new Intent(context, $T.class)", AndroidClassName.INTENT, annotatedTypeName)
                .returns(AndroidClassName.INTENT);

        for (Element e : allFields) {
            String paramName = processor.getParamName(e);
            builder.addParameter(processor.createNullableParam(e, paramName));
        }
        addIntentStatement(builder, allFields);
        builder.addStatement("return intent");

        return builder.build();
    }

    MethodSpec inject() {
        /*
          生成objectHost调用的inject方法
         */
        MethodSpec.Builder builder = MethodSpec.methodBuilder("inject")
                .addModifiers(PUBLIC, STATIC)
                .addParameter(annotatedTypeName, "objectHost")
                .addParameter(AndroidClassName.BUNDLE, "extras");

        builder.beginControlFlow("if (extras == null)")
                .addStatement("return")
                .endControlFlow();

        for (Element e : allFields) {
            String paramName = processor.getParamName(e);
            char[] cs = paramName.toCharArray();

            // 判断是否boolean而且带有is声明, 需要去掉is
            TypeName tn = processor.getTypeName(e);
            if (tn.equals(TypeName.BOOLEAN) && paramName.startsWith("is")) {
                String realName = paramName.substring(2);
                builder.addStatement("objectHost.set$N(($T) extras.get($S))", realName, e.asType(), paramName);
            } else {
                cs[0] -= 32;
                String realName = new String(cs);
                if (tn.box().isBoxedPrimitive()
                        || tn.equals(PrimitiveClassName.STRING)) {
                    // Boolean Byte Short Int Long Char Float Double String Serialize
                    builder.addStatement("objectHost.set$N(($T) extras.get($S))", realName, e.asType(), paramName);
                } else {
                    builder.beginControlFlow("if (extras.containsKey($S))", paramName);

                    builder.addStatement("$T typeToken = new TypeToken<$T>(){}", OtherClassName.TYPE_TOKEN, e.asType());
                    builder.addStatement("$T type = typeToken.getType()", OtherClassName.TYPE);
                    processor.addEson(builder);
                    builder.addStatement("String params = (String) extras.get($S)", paramName);

                    builder.beginControlFlow("try")
                            .addStatement("objectHost.set$N(eson.fromJson(params, type))", realName)
                            .nextControlFlow("catch($T e)", OtherClassName.EXCEPTION)
                            .addStatement("e.printStackTrace()")
                            .endControlFlow();

                    builder.endControlFlow();
                }
            }

        }

        return builder.build();
    }

    private void addIntentStatement(MethodSpec.Builder builder, List<Element> elements) {
        for (Element e : elements) {
            String paramName = processor.getParamName(e);
            builder.beginControlFlow("if ($N != null)", paramName);
            TypeName typeName = processor.getTypeNameBox(e);
            if (typeName.isBoxedPrimitive()
                    || typeName.equals(PrimitiveClassName.STRING)) {
                // Boolean Byte Short Int Long Char Float Double String Serialize
                builder.addStatement("intent.putExtra($S, $N)", paramName, paramName);
            } else {
                processor.addEson(builder);
                builder.beginControlFlow("try");
                builder.addStatement("intent.putExtra($S, eson.toJson($N))", paramName, paramName);
                builder.nextControlFlow("catch($T e)", OtherClassName.EXCEPTION);
                builder.addStatement("e.printStackTrace()");
                builder.endControlFlow();
            }
            builder.endControlFlow();
        }
    }
}

File: code/Kotlin/libs/annotation_processing/kapt/compiler/src/main/java/inject/compiler/method/creator/CommonMethod.kt
package inject.compiler.method.creator

import com.squareup.kotlinpoet.BOOLEAN
import com.squareup.kotlinpoet.ClassName
import com.squareup.kotlinpoet.FunSpec
import com.squareup.kotlinpoet.TypeSpec
import inject.compiler.classname.AndroidClassName
import inject.compiler.classname.OtherClassName
import inject.compiler.method.creator.BaseMethod
import inject.compiler.ext.addEson
import inject.compiler.ext.asTypeName
import inject.compiler.ext.createNonNullContext
import inject.compiler.ext.createNullableParam
import inject.compiler.ext.getParamName
import inject.compiler.ext.isParcelable
import inject.compiler.ext.isPrimitive
import inject.compiler.ext.isSerializable
import inject.compiler.ext.isString
import inject.compiler.processor.FormatSpecifiers
import inject.compiler.processor.CreatorProcessor
import javax.lang.model.element.Element

/**
 * @author yuansui
 * @since 2020/9/8
 */
internal class CommonMethod(
    processor: CreatorProcessor, annotatedElement: Element
) : BaseMethod<CreatorProcessor>(processor, annotatedElement) {

    fun create(fileName: String): TypeSpec {
        val builder = TypeSpec.companionObjectBuilder()

        val clzName = ClassName(processor.getPackageName(annotatedElement), fileName)

        /*
        生成create方法
       */
        val simpleName = clzName.simpleName
        val createBuilder = FunSpec.builder("create")
            .addStatement("val builder = $simpleName()")
            .returns(clzName)

        // 添加必须的参数
        for (e in requireFields) {
            val paramName = e.getParamName()
            createBuilder.addParameter(createNullableParam(e, paramName))
            createBuilder.addStatement(
                "builder.${FormatSpecifiers.NAME} = ${FormatSpecifiers.NAME}",
                paramName,
                paramName
            )
        }
        createBuilder.addStatement("return builder")

        builder.addFunction(createBuilder.build())

        fun inject(): FunSpec {
            val hostName = "objectHost"
            /*
              生成objectHost调用的inject方法
             */
            val builder = FunSpec.builder("inject")
                .addAnnotation(JvmStatic::class) // 为了让java能反射到这个方法
                .addParameter(hostName, annotatedTypeName)
                .addParameter("extras", AndroidClassName.BUNDLE)

            for (e in allFields) {
                val paramName = e.getParamName()

                // 判断是否boolean而且带有is声明, 需要去掉is
                val typeName = e.asTypeName()
                if ((typeName == BOOLEAN || typeName == BOOLEAN.copy(nullable = true)) && paramName.startsWith("is")) {
                    val realName = paramName.substring(2)
                    builder.addStatement(
                        "$hostName.set${FormatSpecifiers.NAME}((${FormatSpecifiers.TYPE}) extras.get(${FormatSpecifiers.STRING}))",
                        realName,
                        typeName,
                        paramName
                    )
                } else {
                    when {
                        typeName.isPrimitive() -> {
                            // Boolean Byte Short Int Long Char Float Double
                            builder.beginControlFlow("if (extras.containsKey(${FormatSpecifiers.STRING}))", paramName)
                                .addStatement(
                                    "extras.get${FormatSpecifiers.TYPE}(${FormatSpecifiers.STRING}).let { $hostName.${FormatSpecifiers.NAME} = it }",
                                    typeName,
                                    paramName,
                                    paramName
                                )
                                .endControlFlow()
                        }

                        typeName.isString() || typeName.isSerializable() || typeName.isParcelable() -> {
                            // String Serialize Parcelable
                            builder.beginControlFlow("if (extras.containsKey(${FormatSpecifiers.STRING}))", paramName)
                                .addStatement(
                                    "extras.get${FormatSpecifiers.TYPE}(${FormatSpecifiers.STRING})?.let { $hostName.${FormatSpecifiers.NAME} = it }",
                                    typeName,
                                    paramName,
                                    paramName
                                )
                                .endControlFlow()
                        }

                        else -> {
                            builder.beginControlFlow("if (extras.containsKey(${FormatSpecifiers.STRING}))", paramName)

                            builder.addStatement(
                                "val typeToken = object : ${FormatSpecifiers.TYPE}<${FormatSpecifiers.TYPE}>(){}",
                                OtherClassName.TYPE_TOKEN,
                                typeName
                            )
                            builder.addStatement("val type = typeToken.getType()", OtherClassName.TYPE)
                            builder.addEson()
                            builder.addStatement("val params = extras.getString(${FormatSpecifiers.STRING})", paramName)

                            builder.beginControlFlow("try")
                                .addStatement(
                                    "eson.fromJson<${FormatSpecifiers.TYPE}>(params, type)?.let { $hostName.${FormatSpecifiers.NAME} = it }",
                                    typeName,
                                    paramName
                                )
                                .nextControlFlow("catch(e : ${FormatSpecifiers.TYPE})", OtherClassName.EXCEPTION)
                                .addStatement("e.printStackTrace()")
                                .endControlFlow()

                            builder.endControlFlow()
                        }
                    }
                }
            }

            return builder.build()
        }
        builder.addFunction(inject())

        return builder.build()
    }

    /**
     * 根据optional生成链式调用的方法
     */
    fun optionalSetters(fileName: String): List<FunSpec> {
        val specs = mutableListOf<FunSpec>()

        for (e in optFields) {
            val paramName = e.getParamName()
            val method = FunSpec.builder(paramName)
                .addParameter(
                    createNullableParam(
                        e,
                        paramName
                    )
                )
                .addStatement("this.${FormatSpecifiers.NAME} = ${FormatSpecifiers.NAME}", paramName, paramName)
                .addStatement("return this")
                .returns(ClassName(processor.getPackageName(annotatedElement), fileName))
                .build()

            specs.add(method)
        }

        return specs
    }

    fun newIntent(): FunSpec {
        val builder = FunSpec.builder("newIntent")
            .addParameter(createNonNullContext())
            .addStatement("val intent = Intent(context, ${FormatSpecifiers.TYPE}::class.java)", annotatedTypeName)
            .returns(AndroidClassName.INTENT)

        for (e in allFields) {
            val paramName = e.getParamName()
            builder.addParameter(createNullableParam(e, paramName))
        }
        addIntentStatement(builder, allFields)
        builder.addStatement("return intent")

        return builder.build()
    }

    private fun addIntentStatement(builder: FunSpec.Builder, elements: List<Element>) {
        for (e in elements) {
            val paramName = e.getParamName()
            builder.beginControlFlow("if (${FormatSpecifiers.NAME} != null)", paramName)
            val typeName = e.asTypeName()
            /**
             * String和Serializable类型单独处理, 因为kotlin的String同时继承了CharSequence和Serializable, putExtra无法识别具体使用哪个
             */
            when {
                typeName.isString() -> {
                    builder.addStatement(
                        "intent.putExtra(${FormatSpecifiers.STRING}, ${FormatSpecifiers.NAME} as CharSequence?)",
                        paramName,
                        paramName
                    )
                }

                typeName.isSerializable() -> {
                    builder.addStatement(
                        "intent.putExtra(${FormatSpecifiers.STRING}, ${FormatSpecifiers.NAME} as ${FormatSpecifiers.TYPE}?)",
                        paramName,
                        paramName
                    )
                }

                typeName.isPrimitive() || typeName.isParcelable() -> {
                    // Boolean Byte Short Int Long Char Float Double String Serialize
                    builder.addStatement(
                        "intent.putExtra(${FormatSpecifiers.STRING}, ${FormatSpecifiers.NAME})",
                        paramName,
                        paramName
                    )
                }

                else -> {
                    builder.addEson()
                    builder.beginControlFlow("try")
                    builder.addStatement(
                        "intent.putExtra(${FormatSpecifiers.STRING}, eson.toJson(${FormatSpecifiers.NAME}))",
                        paramName,
                        paramName
                    )
                    builder.nextControlFlow("catch(e: ${FormatSpecifiers.TYPE})", OtherClassName.EXCEPTION)
                    builder.addStatement("e.printStackTrace()")
                    builder.endControlFlow()
                }
            }
            builder.endControlFlow()
        }
    }
}
